src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -4414,6 +4414,213 @@ class StubGenerator: public StubCodeGenerator {
     return (address) start;
   }
 
+#ifdef COMPILER2
+
+static const int64_t bits2 = right_n_bits(2);
+
+  // In sun.security.util.math.intpoly.IntegerPolynomial1305, integers
+  // are represented as long[5], with BITS_PER_LIMB = 26.
+  // Pack five 26-bit limbs into three 64-bit registers.
+  void poly1305_pack_26(Register dest0, Register dest1, Register dest2, Register src, Register tmp1, Register tmp2) {
+    assert_different_registers(dest0, dest1, dest2, src, tmp1, tmp2);
+
+    // The goal is to have 128-bit value in dest2:dest1:dest0
+    __ ld(dest0, Address(src, 0));    // 26 bits in dest0
+
+    __ ld(tmp1, Address(src, sizeof(jlong)));
+    __ slli(tmp1, tmp1, 26);
+    __ add(dest0, dest0, tmp1);       // 52 bits in dest0
+
+    __ ld(tmp2, Address(src, 2 * sizeof(jlong)));
+    __ slli(tmp1, tmp2, 52);
+    __ add(dest0, dest0, tmp1);       // dest0 is full
+
+    __ srli(dest1, tmp2, 12);         // 14-bit in dest1
+
+    __ ld(tmp1, Address(src, 3 * sizeof(jlong)));
+    __ slli(tmp1, tmp1, 14);
+    __ add(dest1, dest1, tmp1);       // 40-bit in dest1
+
+    __ ld(tmp1, Address(src, 4 * sizeof(jlong)));
+    __ slli(tmp2, tmp1, 40);
+    __ add(dest1, dest1, tmp2);       // dest1 is full
+
+    if (dest2->is_valid()) {
+      __ srli(tmp1, tmp1, 24);
+      __ mv(dest2, tmp1);               // 2 bits in dest2
+    } else {
+#ifdef ASSERT
+      Label OK;
+      __ srli(tmp1, tmp1, 24);
+      __ beq(zr, tmp1, OK);           // 2 bits
+      __ stop("high bits of Poly1305 integer should be zero");
+      __ should_not_reach_here();
+      __ bind(OK);
+#endif
+    }
+  }
+
+  // As above, but return only a 128-bit integer, packed into two
+  // 64-bit registers.
+  void poly1305_pack_26(Register dest0, Register dest1, Register src, Register tmp1, Register tmp2) {
+    poly1305_pack_26(dest0, dest1, noreg, src, tmp1, tmp2);
+  }
+
+  // U_2:U_1:U_0: += (U_2 >> 2) * 5
+  void poly1305_reduce(Register U_2, Register U_1, Register U_0, Register tmp1, Register tmp2) {
+    assert_different_registers(U_2, U_1, U_0, tmp1, tmp2);
+
+    __ srli(tmp1, U_2, 2);
+    __ shadd(tmp1, tmp1, tmp1, tmp2, 2);
```

Comment:
I'm afraid there might be an overflow issue here. 
Consider a smplifed case(8 bits per U): U_2==0b11111111, tmp1 will be 0b111111, tmp1*5 will overflow, and there is carry for it.
If I'm right, then the fix is to use 2 steps rather than one to reduce, please check the aarch64 version for details.

Reply:
Please correct me if I'm wrong:

1. In first loop iteration: after `poly1305_pack_26` before the loop `U_2` is `0b11` max. After any loop iteration the max for `U_2` is `0b100` (as we found out earlier);
2. After `Add U_0 to S_0`: the carry is `0b1` max, plus mandatory `addi` of `0b1` to `S_2` --> `S_2` is `0b110` max;
3. After `wide_mul`'s and `wide_madd`'s: `U_2` is again `0b11` max due to `andi(U_2, R_0, bits2)` (and even `0b0` in case of `0b100` in first step). NB: inside `wide_*` functions `S_2` is unchanged;
4. `mul(U_2, S_2, U_2)`: `U_2` is `0b1111` max --> in `poly1305_reduce` the `tmp1` is `0b11` max, which is safe?

Reply:
Did the above derivation process consider adc(U_2, U_2, U_1HI, t1); at L4564?

Reply:
Oh, I missed it, thanks! However, it doesn't affect much, as far as I can see:
in `adc(U_2, U_2, U_1HI, t1)` register `t1` is `0b1` max, so the max for `U_2` is `0b10000` --> in `poly1305_reduce` the `tmp1` max is `0b100`. Multiplying it by `0b101` we get `0b10100`, which is still ok. Or am I missing something here?

Reply:
I'm bit confused, are there any restrictions on the value of U_1HI?

Reply:
Ohh, it looks like I completely missed the `U_1HI` too, sorry. Let's do a two-step reduce then

Reply:
Done.

Reply:
Thanks for updating.


> If the above issue is not a false one, and the implemention was tested with existing jdk test suite, then I'm afraid we need to add more test to catch this issue.

As we talked above, seems current test did not catch this overlow issue? Could you try to expand the current tests to catch this overflow issue?

Reply:
Hmm, I looked at the code once again and it seems to me that the overflow is actually impossible. Consider the case of 8-bit registers (for simplicity).

1. Above `wide_mul`'s and `wide_madd`'s:
> we know that the top four bits of R_0 and R_1 are zero

  Therefore,`R_1` and `R_0`'s max value (in our case of 8-bit registers) is `0b1111`. Then max value for `RR_1` is `(R_1 >> 2) * 5` = `0b1000`

2. As we figured out earlier: the max value for `S_2` is `0b110`;

Using these facts we can deduce the restrictions for `U_1HI`:
1. `wide_mul(U_1, U_1HI, S_0, R_1)`: let `S_0 = 0b11111111`. Then the `U_1HI` here is `0b1110`;

2. `wide_madd(U_1, U_1HI, S_1, R_0, t1, t2)`: let `S_1 = 0b11111111`, carry is `0b1`, then the `U_1HI` = `U_1HI` + `0b1110` + `0b1` = `0b11101`;

3. `wide_madd(U_1, U_1HI, S_2, RR_1, t1, t2)`: let `S_2 = 0b11111111` carry is `0b1`, then the `U_1HI` = `U_1HI` + `0b111` (we're taking high 8 bits due to `mulhu` instruction) + `0b1` = `0b1000`;

4. `mul(U_2, S_2, U_2)`: max value for `U_2` is `0b10010`;

5. `adc(U_2, U_2, U_1HI, t1)`: max value for `U_2` is `0b10010` + `0b1000` + `0b1` = `0b10100010`;

6. `poly1305_reduce`: max value for `tmp1` is `0b11001000`

Am I missing something once again or is it alright?

Reply:
I'm fine with that.